### PR TITLE
fixed bugs in note selection

### PR DESCRIPTION
**Pull request by Oscar-Pepper**
- fixes bug where underflow can occur on calculating new target value
- fixes bug where largest note is added and new target is not checked before adding extra notes
- fixes bug where a non-dust note is added assuming a grace input is available but could be from wrong pool to single selected note
- fixes bug where non-dust of value 5000 can be selected for dust
- main note selection algorythm now ignores dust to reach target value
- breaks when all unselected ntoes are dust
- better error handling with NonNegativeAmount for balance calculations and appropriate InputSourceError variants
- removes unecessary option and option handling from note transformation allowing for more specific and correct error variant

### CODE REVIEW

File: zingolib/src/wallet/transaction_records_by_id/trait_inputsource.rs [hunk 1]
```diff
@@ -126,31 +124,40 @@ impl InputSource for TransactionRecordsById {
...
+            .collect::<Result<Vec<_>, _>>()
+            .map_err(InputSourceError::InvalidValue)?;
         unselected.sort_by_key(|(_id, value)| *value);
-
         let mut selected = vec![];
-
         let mut index_of_unselected = 0;
+
         loop {
+            // if no unselected notes are available, return the currently selected notes even if the target value has not been reached
```
Comment (zancas) [inline zingolib/src/wallet/transaction_records_by_id/trait_inputsource.rs:138]: `If there are no unselected notes as of this iteration return any previously selected notes even if the target value has not been reached`

File: zingolib/src/wallet/transaction_records_by_id/trait_inputsource.rs [hunk 2]
```diff
@@ -126,31 +124,40 @@ impl InputSource for TransactionRecordsById {
         anchor_height: zcash_primitives::consensus::BlockHeight,
         exclude: &[Self::NoteRef],
     ) -> Result<SpendableNotes<Self::NoteRef>, Self::Error> {
-        let mut unselected =
-            self.get_spendable_note_ids_and_values(sources, anchor_height, exclude);
-
+        let mut unselected = self
+            .get_spendable_note_ids_and_values(sources, anchor_height, exclude)
+            .into_iter()
+            .map(|(id, value)| NonNegativeAmount::from_u64(value).map(|value| (id, value)))
```
Comment (zancas) [inline zingolib/src/wallet/transaction_records_by_id/trait_inputsource.rs:130]: This replaces `u64` with `Zatoshi` unless there's an error.  How about:

`// swap u64s for Zats`

File: zingolib/src/wallet/transaction_records_by_id/trait_inputsource.rs [hunk 3]
```diff
@@ -126,31 +124,40 @@ impl InputSource for TransactionRecordsById {
...
+                break;
+            }
+
+            // update target value for further note selection
+            let selected_notes_total_value = selected
+                .iter()
+                .try_fold(NonNegativeAmount::ZERO, |acc, (_id, value)| acc + *value)
+                .ok_or(InputSourceError::InvalidValue(BalanceError::Overflow))?;
+            let Some(updated_target_value) = target_value - selected_notes_total_value else {
+                // if underflow, target has been reached
```
Comment (zancas) [inline zingolib/src/wallet/transaction_records_by_id/trait_inputsource.rs:149]: The underflow case is represented by `None` here:

https://github.com/zingolabs/librustzcash/blob/4c67b473cf1acc45edf5619e2862f2f3f8775a7a/components/zcash_protocol/src/value.rs#L353


not to be confused with a BalanceError::Underflow...


But the doc-comment for the BalanceError indicates that it's defined for exactly this case:

https://github.com/zingolabs/librustzcash/blob/4c67b473cf1acc45edf5619e2862f2f3f8775a7a/components/zcash_protocol/src/value.rs#L394

This seems inconsistent to me.

File: zingolib/src/wallet/transaction_records_by_id/trait_inputsource.rs [hunk 4]
```diff
@@ -382,41 +402,42 @@ mod tests {
...
-                    AccountId::ZERO,
-                    target_amount,
-                    &[ShieldedProtocol::Sapling, ShieldedProtocol::Orchard],
-                    anchor_height,
-                    &[],
-                ).unwrap();
-            prop_assert_eq!(spendable_notes.sapling().len(), 1);
-            prop_assert_eq!(spendable_notes.orchard().len(), 1);
-        }
+        // TODO: rewrite select_spendable test suite to test a range of cases and target edge cases correctly
```
Comment (zancas) [inline zingolib/src/wallet/transaction_records_by_id/trait_inputsource.rs:405]: I agree!  I think this is high priority..  I suppose we can land your improvements already, but we may as well add sufficient testing.

**Review by zancas — APPROVED** [review]

**Review by zancas — APPROVED** [review]

File: zingolib/src/wallet/transaction_records_by_id/trait_inputsource.rs (hunk 1 above)
Comment (Oscar-Pepper) [reply zingolib/src/wallet/transaction_records_by_id/trait_inputsource.rs:138]: im happy for you to open a PR to make this change, i was just making @fluidvanadium 's comment more correct

File: zingolib/src/wallet/transaction_records_by_id/trait_inputsource.rs (hunk 2 above)
Comment (Oscar-Pepper) [reply zingolib/src/wallet/transaction_records_by_id/trait_inputsource.rs:130]: sure

File: zingolib/src/wallet/transaction_records_by_id/trait_inputsource.rs (hunk 3 above)
Comment (Oscar-Pepper) [reply zingolib/src/wallet/transaction_records_by_id/trait_inputsource.rs:149]: yeah thats correct, as it fails in only one case I don't see it worthy or a PR into LRZ but i agree
